File: leetcode-solutions/ruby/3. Longest Substring Without Repeating Characters.rb
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

# @param {String} s
# @return {Integer}
def length_of_longest_substring(str)
  sub_str = []
  max_sub_str_len = 0
  
  str.each_char do |char|
    char_index = sub_str.index char
    
    if char_index
      sub_str = sub_str[char_index + 1..-1] #remove part from char
      sub_str << char
    else
      sub_str << char
    end
    sub_str_size = sub_str.size
    max_sub_str_len = sub_str_size if max_sub_str_len < sub_str_size
  end
  
  max_sub_str_len
end

File: leetcode-solutions/python/Algorithm-1/3. Longest Substring Without Repeating Characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

from collections import deque

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        q = deque([])
        memo = {}
        max_size = 0
        for _, c in enumerate(s):
            if memo.get(c, 0) > 0:
                # remove chars till duplicate
                while True:
                    temp = q.popleft()
                    memo[temp] -= 1
                    if temp == c:
                        break
                # insert new char
                memo[c] = 1
                q.append(c)
            else:
                memo[c] = 1
                q.append(c)
                max_size = max(max_size, len(q))
        return max_size
